ftw/usermigration/loginnames.py

```diff
@@ -0,0 +1,46 @@
+from Products.CMFCore.utils import getToolByName
+from Products.PluggableAuthService.interfaces.plugins import IUserEnumerationPlugin
+from Products.PluggableAuthService.plugins.ZODBUserManager import IZODBUserManager
+
+
+def migrate_loginnames(context, mapping, mode='move', replace=False):
+    """Migrate Plone loginnames.
+    """
+
+    # Statistics
+    moved = []
+    copied = []
+    deleted = []
+
+    uf = getToolByName(context, 'acl_users')
+
+    for old_userid, new_userid in mapping.items():
+        for plugin_id, plugin in uf.plugins.listPlugins(IUserEnumerationPlugin):
+            # Only ZODB User Manager is supported
+            if not IZODBUserManager.providedBy(plugin):
+                continue
+
+            for user in plugin.enumerateUsers(id=old_userid, exact_match=True):
```

Comment:
As discussed, you would have to track a separate login-name table to do a login-name migration. You also wont be able to do a login-name and user-id migration at once if the userid differs to the login-name.

i.e.:
userid: chuck
loginname: chuckN

migration-mapping:
chuck: chuck@norris.ch

if you do a loginname and userid migration at once then I would except to get the following result:

userid: chuck@norris.ch
loginname: chuck@norris.ch

But with your requested changes the result would be:

userid: chuck@norris.ch
loginname: chuckN

and you would have to do another login-name migration with the following mapping:

chuckN: chuck@norris.ch